src/ecs/resources/input.rs

```diff
@@ -38,6 +51,112 @@ impl<'a> Iterator for PressedMouseButtons<'a> {
+}
+
+impl<'a> Iterator for PressedButtons<'a> {
+    type Item = Button;
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iterator.next()
+    }
+}
+
+pub struct Axis {
+    pub pos: Button,
+    pub neg: Button,
+}
+
+#[derive(Default)]
+struct FrameInputData {
+    pressed_keys: Vec<VirtualKeyCode>,
+    pressed_mouse_buttons: Vec<MouseButton>,
+    mouse_position: Option<(i32, i32)>,
+}
+
+impl FrameInputData {
+    pub fn new() -> FrameInputData {
+        FrameInputData {
+            pressed_keys: Vec::new(),
+            pressed_mouse_buttons: Vec::new(),
+            mouse_position: None,
+        }
+    }
+
+    fn advance_frame(&mut self, previous_frame: &mut FrameInputData) {
+        previous_frame.mouse_position = self.mouse_position;
+        previous_frame.pressed_keys.clear();
+        previous_frame
+            .pressed_keys
+            .extend_from_slice(&self.pressed_keys);
+        previous_frame.pressed_mouse_buttons.clear();
+        previous_frame
+            .pressed_mouse_buttons
+            .extend_from_slice(&self.pressed_mouse_buttons);
+    }
+
+    /// Returns an iterator for all the pressed down keys.
+    fn pressed_keys(&self) -> PressedKeys {
+        PressedKeys { iterator: self.pressed_keys.iter() }
+    }
+
+    /// Checks if the given key is being pressed.
+    fn key_pressed(&self, key: VirtualKeyCode) -> bool {
+        self.pressed_keys.iter().any(|&k| k == key)
+    }
+
+    /// Checks if all the given keys are being pressed.
+    fn keys_pressed(&self, keys: &[VirtualKeyCode]) -> bool {
+        keys.iter().all(|key| self.key_pressed(*key))
+    }
+
+    /// Returns an iterator for all the pressed down mouse buttons.
+    fn pressed_mouse_buttons(&self) -> PressedMouseButtons {
+        PressedMouseButtons { iterator: self.pressed_mouse_buttons.iter() }
+    }
+
+    /// Checks if the given mouse button is being pressed.
+    fn mouse_button_pressed(&self, button: MouseButton) -> bool {
+        self.pressed_mouse_buttons.iter().any(|&b| b == button)
+    }
+
+    /// Checks if all the given mouse buttons are being pressed.
+    fn mouse_buttons_pressed(&self, buttons: &[MouseButton]) -> bool {
+        buttons.iter().all(|btn| self.mouse_button_pressed(*btn))
+    }
+
+    /// Gets the current mouse position.
+    ///
+    /// this method can return None, either if no mouse is connected, or if no mouse events have
+    /// been recorded
+    fn mouse_position(&self) -> Option<(i32, i32)> {
+        self.mouse_position
+    }
+
+    /// Returns an iterator over the buttons that are currently pressed
+    fn pressed_buttons(&self) -> PressedButtons {
+        let mouse_button_convert = mouse_button_to_button as fn(&MouseButton) -> Button;
+        let key_convert = key_to_button as fn(&VirtualKeyCode) -> Button;
+        let mouse_buttons = self.pressed_mouse_buttons().map(mouse_button_convert);
+        let keys = self.pressed_keys().map(key_convert);
```

Comment:
I sincerely would love to do this, but this is the compiler error I get:

```
jacob@jacob-laptop ~/amethyst $ cargo check
   Compiling amethyst v0.4.3 (file:///home/jacob/amethyst)
error: non-capturing closure to fn coercion is experimental (see issue #39817)
   --> src/ecs/resources/input.rs:139:36
    |
139 |         let mouse_button_convert = (|&mb| Button::Mouse(mb)) as fn(&MouseButton) -> Button;
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^

error: non-scalar cast: `[closure@src/ecs/resources/input.rs:139:36: 139:61]` as `fn(&glutin::MouseButton) -> ecs::resources::input::Button`
   --> src/ecs/resources/input.rs:139:36
    |
139 |         let mouse_button_convert = (|&mb| Button::Mouse(mb)) as fn(&MouseButton) -> Button;
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: non-capturing closure to fn coercion is experimental (see issue #39817)
   --> src/ecs/resources/input.rs:140:27
    |
140 |         let key_convert = (|&k| Button::Key(k)) as fn(&VirtualKeyCode) -> Button;
    |                           ^^^^^^^^^^^^^^^^^^^^^

error: non-scalar cast: `[closure@src/ecs/resources/input.rs:140:27: 140:48]` as `fn(&glutin::VirtualKeyCode) -> ecs::resources::input::Button`
   --> src/ecs/resources/input.rs:140:27
    |
140 |         let key_convert = (|&k| Button::Key(k)) as fn(&VirtualKeyCode) -> Button;
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

error: Could not compile `amethyst`.

To learn more, run the command again with --verbose.
```

These `fn`s have to be stored in the PressedButtons iterator that is passed back so it is necessary to define a type for them.  I could just Box the iterator but then we're moving this structure onto the heap unnecessarily.

Basically this whole system ends up being a type system jigsaw puzzle and this was the only way I could see to solve it and maintain performance.